File: internal/notifications/tool_maps.go
// tool_maps.go.
// Builds tool_use and tool_result index maps from a parsed message slice.
package notifications

import (
	"encoding/json"

	"claude-devtools/internal/domain"
)

// ToolUseInfo mirrors trigger_checker/tool_maps.rs ToolUseInfo.
type ToolUseInfo struct {
	Name string
}

// ToolResultInfo mirrors trigger_checker/tool_maps.rs ToolResultInfo.
type ToolResultInfo struct {
	Content string
}

// BuildToolUseMap indexes tool_use blocks by ID across assistant messages.
// Mirrors trigger_checker/tool_maps.rs::build_tool_use_map.
func BuildToolUseMap(messages []domain.ParsedMessage) map[string]ToolUseInfo {
	m := make(map[string]ToolUseInfo)
	for i := range messages {
		msg := &messages[i]
		if msg.MessageType != "assistant" {
			continue
		}
		for j := range msg.Content.Blocks {
			b := &msg.Content.Blocks[j]
			if b.Type == "tool_use" && b.ID != nil && b.Name != nil {
				m[*b.ID] = ToolUseInfo{Name: *b.Name}
			}
		}
		for j := range msg.ToolCalls {
			tc := &msg.ToolCalls[j]
			if _, exists := m[tc.ID]; !exists {
				m[tc.ID] = ToolUseInfo{Name: tc.Name}
			}
		}
	}
	return m
}

// BuildToolResultMap indexes tool_result content by tool_use ID.
// Mirrors trigger_checker/tool_maps.rs::build_tool_result_map.
func BuildToolResultMap(messages []domain.ParsedMessage) map[string]ToolResultInfo {
	m := make(map[string]ToolResultInfo)
	for i := range messages {
		msg := &messages[i]
		for j := range msg.Content.Blocks {
			b := &msg.Content.Blocks[j]
			if b.Type == "tool_result" && b.ToolUseID != nil && b.Content != nil {
				content := toolResultContentValue(b.Content)
				if _, exists := m[*b.ToolUseID]; !exists {
					m[*b.ToolUseID] = ToolResultInfo{Content: content}
				}
			}
		}
		for j := range msg.ToolResults {
			tr := &msg.ToolResults[j]
			if _, exists := m[tr.ToolUseID]; !exists {
				m[tr.ToolUseID] = ToolResultInfo{Content: valueToContentString(tr.Content)}
			}
		}
		if msg.ToolUseResult != nil && msg.SourceToolUseID != nil {
			content := extractContentFromToolUseResult(msg.ToolUseResult)
			if _, exists := m[*msg.SourceToolUseID]; !exists {
				m[*msg.SourceToolUseID] = ToolResultInfo{Content: content}
			}
		}
	}
	return m
}

// toolResultContentValue extracts string content from a ToolResultContentValue.
func toolResultContentValue(v *domain.ToolResultContentValue) string {
	if v.Text != nil {
		return *v.Text
	}
	return extractTextFromBlocks(v.Blocks)
}

// extractTextFromBlocks joins text blocks with newlines.
// Mirrors trigger_checker/tool_maps.rs::extract_text_from_blocks.
func extractTextFromBlocks(blocks []domain.ContentBlock) string {
	var parts []string
	for i := range blocks {
		b := &blocks[i]
		if b.Type == "text" && b.Text != nil {
			parts = append(parts, *b.Text)
		}
	}
	result := ""
	for i, p := range parts {
		if i > 0 {
			result += "\n"
		}
		result += p
	}
	return result
}

// valueToContentString converts a json.RawMessage to string.
// Mirrors trigger_checker/tool_maps.rs::value_to_content_string.
func valueToContentString(raw domain.RawValue) string {
	if raw == nil {
		return ""
	}
	var s string
	if json.Unmarshal(raw, &s) == nil {
		return s
	}
	return string(raw)
}

// extractContentFromToolUseResult extracts the most useful text from a tool_use_result JSON object.
// Mirrors trigger_checker/tool_maps.rs::extract_content_from_tool_use_result.
func extractContentFromToolUseResult(raw domain.RawValue) string {
	if raw == nil {
		return ""
	}
	var obj map[string]json.RawMessage
	if json.Unmarshal(raw, &obj) != nil {
		return ""
	}
	getStr := func(key string) string {
		v, ok := obj[key]
		if !ok {
			return ""
		}
		var s string
		if json.Unmarshal(v, &s) == nil {
			return s
		}
		return ""
	}
	if s := getStr("error"); s != "" {
		return s
	}
	if s := getStr("stderr"); s != "" {
		return s
	}
	if s := getStr("content"); s != "" {
		return s
	}
	return getStr("message")
}

File: src-tauri/src/notifications/tool_maps.rs
//! Builds tool_use / tool_result index maps from a parsed message slice.
//! Ported from `internal/notifications/tool_maps.go`.

use std::collections::HashMap;

use serde_json::Value;

use crate::types::jsonl::{ContentBlock, ToolResultContentValue};
use crate::types::messages::{ParsedMessage, ParsedMessageContent};

#[derive(Debug, Clone)]
pub struct ToolUseInfo {
    pub name: String,
}

#[derive(Debug, Clone)]
pub struct ToolResultInfo {
    pub content: String,
}

/// Content blocks of a message, or an empty slice when the content is text.
pub(crate) fn message_blocks(msg: &ParsedMessage) -> &[ContentBlock] {
    match &msg.content {
        ParsedMessageContent::Blocks(blocks) => blocks,
        ParsedMessageContent::Text(_) => &[],
    }
}

/// Indexes tool_use blocks by ID across assistant messages.
pub fn build_tool_use_map(messages: &[ParsedMessage]) -> HashMap<String, ToolUseInfo> {
    let mut map = HashMap::new();
    for msg in messages {
        if msg.message_type != "assistant" {
            continue;
        }
        for block in message_blocks(msg) {
            if let ContentBlock::ToolUse { id, name, .. } = block {
                map.insert(id.clone(), ToolUseInfo { name: name.clone() });
            }
        }
        for tc in &msg.tool_calls {
            map.entry(tc.id.clone())
                .or_insert_with(|| ToolUseInfo { name: tc.name.clone() });
        }
    }
    map
}

/// Indexes tool_result content by tool_use ID.
pub fn build_tool_result_map(messages: &[ParsedMessage]) -> HashMap<String, ToolResultInfo> {
    let mut map = HashMap::new();
    for msg in messages {
        for block in message_blocks(msg) {
            if let ContentBlock::ToolResult {
                tool_use_id,
                content,
                ..
            } = block
            {
                let value = tool_result_content_value(content);
                map.entry(tool_use_id.clone())
                    .or_insert_with(|| ToolResultInfo { content: value });
            }
        }
        for tr in &msg.tool_results {
            let value = value_to_content_string(&tr.content);
            map.entry(tr.tool_use_id.clone())
                .or_insert_with(|| ToolResultInfo { content: value });
        }
        if let (Some(tur), Some(source_id)) = (&msg.tool_use_result, &msg.source_tool_use_id) {
            let value = extract_content_from_tool_use_result(tur);
            map.entry(source_id.clone())
                .or_insert_with(|| ToolResultInfo { content: value });
        }
    }
    map
}

/// Extracts string content from a tool_result content value.
pub(crate) fn tool_result_content_value(v: &ToolResultContentValue) -> String {
    match v {
        ToolResultContentValue::Text(s) => s.clone(),
        ToolResultContentValue::Blocks(blocks) => extract_text_from_blocks(blocks),
    }
}

/// Joins text blocks with newlines.
fn extract_text_from_blocks(blocks: &[ContentBlock]) -> String {
    blocks
        .iter()
        .filter_map(|b| match b {
            ContentBlock::Text { text } => Some(text.as_str()),
            _ => None,
        })
        .collect::<Vec<_>>()
        .join("\n")
}

/// Converts a JSON value to a string: strings unwrap, null → "", else JSON text.
pub(crate) fn value_to_content_string(v: &Value) -> String {
    match v {
        Value::Null => String::new(),
        Value::String(s) => s.clone(),
        other => other.to_string(),
    }
}

/// Extracts the most useful text from a tool_use_result JSON object.
pub(crate) fn extract_content_from_tool_use_result(v: &Value) -> String {
    let obj = match v.as_object() {
        Some(o) => o,
        None => return String::new(),
    };
    let get_str = |key: &str| -> String {
        obj.get(key)
            .and_then(|x| x.as_str())
            .unwrap_or("")
            .to_string()
    };
    for key in ["error", "stderr", "content"] {
        let s = get_str(key);
        if !s.is_empty() {
            return s;
        }
    }
    get_str("message")
}

#[cfg(test)]
#[path = "tool_maps_tests.rs"]
mod tests;
